developer-docs/kerberos.md

```diff
@@ -0,0 +1,73 @@
+# Bolt + Kerberos
+
+### Overview
+
+Bolt supports using Kerberos as an authentication mechanism, in lieu of a username / password. An excellent background on key Kerberos concepts is available in [Designing an Authentication System: a Dialogue in Four Scenes](https://web.mit.edu/kerberos/www/dialogue.html).
+
+While Linux **does** support a standalone Kerberos KDC (Key Distribution Center), the [`winrm`](https://github.com/WinRb/WinRM) gem that provides connectivity support from Bolt has really only been written for / tested against Active Directory.
+
+In the future, it's possible other transports like SSH may be able to authenticate and authorize with just a KDC, but for now, the primary use case for Kerberos is in conjunction with WinRM.
+
+### Kerberos Implementations
+
+There are three primary implementations of the Kerberos protocol widely available:
+
+* [Heimdal](#heimdal)
+* [MIT Kerberos](#mit-kerberos)
+* [Microsoft Kerberos](#microsoft-kerberos)
+
+#### Heimdal
+
+[Heimdal](https://www.h5l.org/) is an open source implementation that is shipped on Mac OSX and is the default library used for the Samba server packages on Linux.
+
+Even though Heimdal provided interoperability with [Microsoft DCE/RPC](https://en.wikipedia.org/wiki/DCE/RPC) first with the addition of [IOV message wrapping extension functions](https://web.mit.edu/kerberos/krb5-latest/doc/appdev/gssapi.html#iov-message-wrapping), these APIs are not exported in the OSX libraries that ship with the operating system.
+
+The [gssapi](https://github.com/zenchild/gssapi) gem that Bolt relies on for [Generic Security Services](https://en.wikipedia.org/wiki/Generic_Security_Services_Application_Program_Interface) defaults to MIT Kerberos, but also supports Heimdal via a programmatic opt-in by adding `require 'gssapi/heimdal'` prior to `require 'gssapi'` (for instance, to the code in `winrm/connection.rb`). Unfortunately the support for Heimdal is not well tested, the paths to the Heimdal libraries are currently hard-coded, and the library doesn't account for semantic differences in the Heimdal and MIT implementations of Kerberos APIs, which can cause segfaults.
```

Comment:
Is the TL;DR here that Heimdal isn't supported?

Reply:
It's a bit subtle and I tried explaining in as few words as possible.

`gssapi` supports Heimdal... under specific circumstances.

But Bolt doesn't opt-in under the circumstances where it might work anyhow (though could if we wanted to test it, etc).

Reply:
I see so you could patch Bolt to work with heimdal? But not use it out of the box?

Reply:
Heimdal is a cross-platform library... and `gssapi` loads it differently based on platform:
https://github.com/zenchild/gssapi/blob/826c91b8481407b0db36cb97fe88f49ac3e21315/lib/gssapi/lib_gssapi_loader.rb#L46-L57

I only tested on OSX... to even get it to load, I had to:

* put a `require 'gssapi/heimdal'` before other `gssapi` requires in the Bolt source
* patch the above load path (since I was trying to use system Heimdal, not a separately installed Heimdal)
* patch parts of the `gssapi` gem itself to prevent two different double-free segfaults

Even then, OSX system Heimdal doesn't export the `IOV` methods we care about... so it was a dead end basically.

I didn't try:

* using Heimdal on Linux
* using a different Heimdal binary on OSX

Plausibly those work without patches, but I didn't try because:

* Linux MIT Kerberos is the most prevalent (except inside Samba server, go figure)
* I filed a separate ticket to figure out how we'll address the OSX situation... we could compile + ship MIT or Heimdal binaries with the OSX packages, for instance